web_client/views/VideoWidget.js

```diff
@@ -0,0 +1,92 @@
+import $ from 'jquery';
+import _ from 'underscore';
+import reconcile from 'reconcile.js';
+import View from 'girder/views/View';
+import { getApiRoot, restRequest } from 'girder/rest';
+
+import ModalView from './ModalView';
+
+import template from '../templates/VideoWidget.pug';
+
+import '../stylesheets/VideoWidget.styl';
+
+var VideoWidget = View.extend({
+    events: {
+        'click button.convert': function (e) {
+            this.status = 2;
+            this.playing = false;
+            this.render();
+            setTimeout(() => {
+                this.status = 3;
+                this.render();
+            }, 3000);
+        },
+        'click button.delete': function (e) {
+            this.status = 0;
+            this.playing = false;
+            this.render();
+        },
+        'click button.play': function (e) {
+            this.videoElement.play();
+        },
+        'click button.pause': function (e) {
+            this.videoElement.pause();
+        },
+        'click button.popup': function (e) {
+            new ModalView({
+                el: $('#g-dialog-container'),
+                model: this.fileModel,
+                parentView: this
+            }).render();
+        }
+    },
+    initialize(settings) {
+        this.fileModel = settings.model;
+        this.modalMode = settings.modalMode;
+        this.status = (() => {
+            let status = Math.floor(Math.random() * 3);
+            if (status !== 0) {
+                status += 1;
+            }
+            return status;
+        })();
+        if (this.status === 2) {
+            setTimeout(() => {
+                this.status = 3;
+                this.render();
+            }, 3000);
+        }
+    },
+    render() {
+        let getTemplate = () => {
+            return template(Object.assign({ src: `${getApiRoot()}/file/${this.fileModel.id}/download` }, this));
+        };
+        if (!this.initialized) {
+            this.initialized = true;
+            this.$el.html(getTemplate());
+        }
+        else {
+            this.update(getTemplate());
+        }
+        if (this.videoElement) {
+            $(this.videoElement).off();
+        }
+        this.videoElement = this.$('.g-video-preview')[0];
+        this.$('.g-video-preview').on('play', () => {
+            this.playing = true;
+            this.render();
+        });
+        this.$('.g-video-preview').on('pause', () => {
+            this.playing = false;
+            this.render();
+        });
+        return this;
+    },
+    update(template) {
+        var $new = $(template);
+        var changes = reconcile.diff($new[0], this.$el.children(0)[0]);
```

Comment:
These two lines are confusing to read through, it's hard to know what this means

    $new[0], this.$el.children(0)[0]

Perhaps some well named variables or code comments would help with readability.

Reply:
This is related to the reconcile DOM update, in Minerva, I put this update() function on the base View class. Basically, it updates the DOM without destroying the old one. I could create a base class or utility class for this.
Generally, I created PR as a mock-up to get things started. I was thinking all the code here won't stay for long.